### DISCUSSION TITLE
null value in RecursiveItems

### DISCUSSION THREAD

**Discussion by bark92** [halaxa/json-machine#119] [General]
some value of my json is contain value null. is there any method to use RecursiveItems when have null as value.?
for example :
```json
[
  {
    "username": "user",
    "e-mail": "user@example.com",
    "friends": [
      {
        "username": "friend1",
        "e-mail": "friend1@example.com"
      },
      {
        "username": "friend2",
        "e-mail": null
      }
    ]
  }
]
```

**Comment by halaxa** [comment]
Can you be more specific? What code does not work for you?

**Reply by bark92** [reply]
this is my script : 
```php
<?php
use JsonMachine\RecursiveItems;
$users = RecursiveItems::fromFile(basedir.'/public/tmp/test.json');
foreach ($users as $user) {
    /** @var $user RecursiveItems */
    foreach ($user as $field => $value) {
        if ($field === 'friends') {
            /** @var $value RecursiveItems */
            foreach ($value as $friend) {
                /** @var $friend RecursiveItems */
                $data = $friend->toArray();
                print_r($data);
            }
        }
    }
}
```
and this is my json structure : 
```json
[
    {
      "username": "user",
      "e-mail": "user@example.com",
      "friends": [
        {
          "username": "friend1",
          "e-mail": "friend1@example.com"
        },
        {
          "username": "friend2",
          "e-mail": null
        }
      ]
    }
  ]
```
this the result : 

Array ( [username] => friend1 [e-mail] => friend1@example.com )
Fatal error: Uncaught JsonMachine\Exception\InvalidArgumentException: JsonMachine\RecursiveItems only accepts scalar or IteratorAggregate values. NULL given. in C:\laragon\www\library\vendor\halaxa\json-machine\src\RecursiveItems.php:105 Stack trace: #0 C:\laragon\www\library\vendor\halaxa\json-machine\src\RecursiveItems.php(210): JsonMachine\RecursiveItems->current() #1 C:\laragon\www\library\vendor\halaxa\json-machine\src\RecursiveItems.php(204): JsonMachine\RecursiveItems::toArrayRecursive(Object(JsonMachine\RecursiveItems)) #2 C:\laragon\www\controller\global\testjson.php(11): JsonMachine\RecursiveItems->toArray() #3 {main} thrown in C:\laragon\www\library\vendor\halaxa\json-machine\src\RecursiveItems.php on line 105

**Reply by halaxa** [reply]
You might have found a bug. I'll look into it. Thanks

**Reply by bark92** [reply]
i modify this function in RecursiveItems.php:
`public function current()
    {
        $current = $this->parserIterator->current();
        if ($current instanceof IteratorAggregate) {
            return new self($current, $this->options);
        } elseif ( ! is_scalar($current)) {
            throw new InvalidArgumentException(
                sprintf(
                    '%s only accepts scalar or IteratorAggregate values. %s given.',
                    self::class,
                    is_object($current) ? get_class($current) : gettype($current)
                )
            );
        }
        return $current;
    }`
in to this : 
`public function current()
    {
        $current = $this->parserIterator->current();
        if ($current === null) {
            return null;
        } elseif ($current instanceof IteratorAggregate) {
            return new self($current, $this->options);
        } elseif (!is_scalar($current)) {
            throw new InvalidArgumentException(
                sprintf(
                    '%s only accepts scalar or IteratorAggregate values. %s given.',
                    self::class,
                    is_object($current) ? get_class($current) : gettype($current)
                )
            );
        }
        return $current;
    }` 
it fixed my problem. and everything just run smooth. even i set my memory limit in to 256, i can parse my 1gb json successfully without memory exhausted issue. thank you so much

**Comment by Vanilagy** [comment]
Yeah I also hit this. @bark92's fix worked. More specifically, the function looks like this now:
```ts
/**
     * @return mixed Move to return type when PHP 7 support is dropped
     */
    #[\ReturnTypeWillChange]
    public function current()
    {
        $current = $this->parserIterator->current();
        if ($current instanceof IteratorAggregate) {
            return new self($current, $this->options);
        } elseif ( ! is_scalar($current) && $current !== null) {
            throw new InvalidArgumentException(
                sprintf(
                    '%s only accepts scalar or IteratorAggregate values. %s given.',
                    self::class,
                    is_object($current) ? get_class($current) : gettype($current)
                )
            );
        }

        return $current;
    }
```

Keep in mind that `is_scalar(null)` returns false in PHP.

**Comment by halaxa** [comment]
Thanks. Fixed here 52685e11d8e053802a730c95a9e29fa88cc1aa44

**Reply by Vanilagy** [reply]
Awesome, thank you!

**Reply by Vanilagy** [reply]
When will you be publishing a new version?

**Reply by halaxa** [reply]
Released 1.2.1. Packagist may take a few minutes to notice.

**Reply by Vanilagy** [reply]
Arigatou!
